openfeature/api.go

```diff
@@ -166,32 +216,35 @@ func (api *evaluationAPI) forTransaction(clientName string) (FeatureProvider, []
 	return provider, api.hks, api.evalCtx
 }
 
-// initNewAndShutdownOld is a helper to initialise new FeatureProvider and shutdown the old FeatureProvider.
+// initAndNotify is a helper to initialise FeatureProvider and notify the event executor.
+func initAndNotify(provider FeatureProvider, stateHandler StateHandler, evalCtx EvaluationContext, eventChan chan eventPayload) {
+	err := stateHandler.Init(evalCtx)
+	// emit ready/error event once initialization is complete
+	if err != nil {
+		eventChan <- eventPayload{
+			Event{
+				ProviderName:         provider.Metadata().Name,
+				EventType:            ProviderError,
+				ProviderEventDetails: ProviderEventDetails{},
+			}, provider,
+		}
+	} else {
+		eventChan <- eventPayload{
+			Event{
+				ProviderName:         provider.Metadata().Name,
+				EventType:            ProviderReady,
+				ProviderEventDetails: ProviderEventDetails{},
+			}, provider,
+		}
+	}
+}
+
+// initNewAndShutdownOldAsync is a helper to initialise new FeatureProvider and shutdown the old FeatureProvider.
 // Operations happen concurrently.
-func (api *evaluationAPI) initNewAndShutdownOld(newProvider FeatureProvider, oldProvider FeatureProvider) {
+func (api *evaluationAPI) initNewAndShutdownOldAsync(newProvider FeatureProvider, oldProvider FeatureProvider) {
```

Comment:
You may refactor to combine `initNewAndShutdownOldAsync` & `initNewAndShutdownOldSync` as they perform similar set of operations with only difference being sync vs async.